### DISCUSSION TITLE
Technical changes: Delete and concurrency rework

### DISCUSSION THREAD

**Discussion by AlUlkesh** [AlUlkesh/stable-diffusion-webui-images-browser#94] [Announcements]
This started with trying to fix the delete functionality, which had been bandaid-fixed several times and still didn't work right in difficult to reproduce circumstances. And deleting the wrong file is always very bad.

While deep in debugging I noticed more problems with the code, especially related to synchronisation with Gradio, but also javascript functions running at the same time, that should not do that.

What's more, while fixing bit by bit, other parts started failing, that apparently depended on the other wrong code. I suspect these were put in as workarounds at some point.

After many many different attempts, I decided to completely rework the whole logic.

This current version seems very stable and responsive in my tests. Also it will make future changes much easier for me.

Some more details about what was changed:
- Delete logic changed from "guessing and hiding" in-memory objects to actually removing them
- Remove timeouts/waits
- Handle concurrency with locks, promises and async/await
- Correct page index calculation
- Rework python and javascript dependencies, remove randomness
- Prevent errors caused by "button spamming"
- Prevent hundreds of EventListeners being added again and again without removing them

**Comment by AlUlkesh** [comment]
https://github.com/AlUlkesh/stable-diffusion-webui-images-browser/commit/62cf141330d932f6d49ef4a32891f4c17636e585

**Comment by Aamir3d** [comment]
Great going, thanks! Will check this out and update.
